File: Algorithms/Python/0019/0019.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummyHead = ListNode(0)
        dummyHead.next = head
        slow, fast = dummyHead, dummyHead
        for i in range(n):
            fast = fast.next
        while fast.next:
            slow, fast = slow.next, fast.next
        slow.next = slow.next.next
        return dummyHead.next

File: Algorithms/C++/0019/0019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *dummyHead = new ListNode(0);
        dummyHead->next = head;
        ListNode *p = dummyHead, *q = dummyHead;
        for (int i = 0; i < n; i++) {
            q = q->next;
        }
        while (q->next) {
            p = p->next;
            q = q->next;
        }
        p->next = p->next->next;
        return dummyHead->next;
    }
};
